### PR TITLE
Ability to set the schema pattern for tables metadata retrieval

### PR DESCRIPTION
**Pull request by pqueixalos**
This adds the ability to configure jdbc schemaPattern from connector configuration.
Reported here https://github.com/confluentinc/kafka-connect-jdbc/issues/32

### CODE REVIEW

**Comment by pqueixalos** [comment]
CLA just signed.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 1]
```diff
@@ -144,6 +143,9 @@
   public static final long TIMESTAMP_DELAY_INTERVAL_MS_DEFAULT = 0;
   private static final String TIMESTAMP_DELAY_INTERVAL_MS_DISPLAY = "Delay Interval (ms)";
 
+  public static final String SCHEMA_PATTERN_CONFIG = "schemaPattern";
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: `schema.pattern` to be consistent with the config naming conventions

File: src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingTableQuerier.java [hunk 2]
```diff
@@ -157,9 +155,9 @@ protected ResultSet executeQuery() throws SQLException {
       stmt.setLong(3, incOffset);
       stmt.setTimestamp(4, tsOffset, UTC_CALENDAR);
       log.debug("Executing prepared statement with start time value = {} end time = {} and incrementing value = {}",
-              JdbcUtils.formatUTC(tsOffset),
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingTableQuerier.java]: minor: can you please remove the whitespace changes

**Comment by shikhar** [comment]
@pqueixalos is it possible to just include the schema in the connection URL?

**Comment by pqueixalos** [comment]
@shikhar, short answer : no. 
Settting the schema in URL configures the JDBC Driver and not all JDBC implementations supports this.
Also it would not change anything (I tried :)), the main issue is that JdbcUtils is aksing for database metadata without providing any schema pattern so every tables of every 'usable' schemas are retrieved.

Regarding the changes you asked (whitespace and parameter naming, sure, I will make those changes next week).

**Comment by shikhar** [comment]
@pqueixalos would it work to update the code to use `connection.getSchema()` as the schema pattern? (in conjunction with the user specifying it as part of the connection URL)

File: src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java [hunk 3]
```diff
@@ -16,6 +16,12 @@
 
 package io.confluent.connect.jdbc;
 
+import io.confluent.connect.jdbc.source.JdbcSourceConnectorConfig;
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java]: let's avoid the import rearrangement

**Comment by pqueixalos** [comment]
Good point @shikhar, getting the schema from the connection (when set in jdbc connection params) might work fine in some situations !
It won't work in my context for sure, as (silly) Redshift connector does not support it. Also, I don't know how much other jdbc connector implementations honor this, for instance MySQL would probably return null as it does not implement schema concept, connection.getCatalog() would need to be used instead.

**Comment by pqueixalos** [comment]
[clabot:check]

File: src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java [hunk 4]
```diff
@@ -104,8 +105,8 @@ public void start(Map<String, String> properties) throws ConnectException {
       // query.
       whitelistSet = Collections.emptySet();
     }
-    tableMonitorThread = new TableMonitorThread(db, context, tablePollMs, whitelistSet,
-                                                blacklistSet, tableTypesSet);
+    tableMonitorThread = new TableMonitorThread(db, schemaPattern, context, tablePollMs,
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java]: @pqueixalos sorry didn't notice this minor thing previously, but in terms of arg ordering `schemaPattern` makes sense after the context

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 5]
```diff
@@ -73,6 +74,7 @@ public void start(Map<String, String> properties) {
       throw new ConnectException("Couldn't start JdbcSourceTask due to configuration error", e);
     }
 
+    this.schemaPattern = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: I don't think we need this as a field on this class.

**Comment by shikhar** [comment]
@pqueixalos ah too bad Redshift JDBC does not support that. So we need this patch to make the JDBC source connector usable against Redshift then.

Left a couple more comments, otherwise LGTM

cc @ewencp in case you have any thoughts on this

**Comment by shikhar** [comment]
ok to test

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 6]
```diff
@@ -190,6 +192,7 @@ public static ConfigDef baseConfigDef() {
         .define(BATCH_MAX_ROWS_CONFIG, Type.INT, BATCH_MAX_ROWS_DEFAULT, Importance.LOW, BATCH_MAX_ROWS_DOC, CONNECTOR_GROUP, 2, Width.SHORT, BATCH_MAX_ROWS_DISPLAY)
         .define(TABLE_POLL_INTERVAL_MS_CONFIG, Type.LONG, TABLE_POLL_INTERVAL_MS_DEFAULT, Importance.LOW, TABLE_POLL_INTERVAL_MS_DOC, CONNECTOR_GROUP, 3, Width.SHORT, TABLE_POLL_INTERVAL_MS_DISPLAY)
         .define(TOPIC_PREFIX_CONFIG, Type.STRING, Importance.HIGH, TOPIC_PREFIX_DOC, CONNECTOR_GROUP, 4, Width.MEDIUM, TOPIC_PREFIX_DISPLAY)
+        .define(SCHEMA_PATTERN_CONFIG, Type.STRING, null, Importance.MEDIUM, SCHEMA_PATTERN_DOC, DATABASE_GROUP, 5, Width.SHORT, QUERY_DISPLAY)
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: Is this in the right location? If it is part of the `DATABASE_GROUP` should we relocate it with that group? And I think the order might not be quite right since it looks like there are only 3 items in the `DATABASE_GROUP` so far. (It'd also be nice to include a DISPLAY setting as well for a more human-friendly form.)

**Comment by ewencp** [comment]
@pqueixalos @shikhar Only took a quick scan through, but the changes seem fine. Major missing piece seems to be tests -- there were some updates to tests but it looks like they just use `null` for the schema pattern so we haven't actually tested the case where the user sets the schema pattern.

**Comment by pqueixalos** [comment]
@ewencp I added a test in `JdbcUtilsTest`, but taht is probably not enough. Making sure that the configuration is well handled at `JdbcSourceConnector` level would be better. I will take care of this on a later commit.
I also improved the configuration documentation as you suggested, hoping I did well.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 7]
```diff
@@ -108,6 +108,14 @@
   public static final String TABLE_BLACKLIST_DEFAULT = "";
   private static final String TABLE_BLACKLIST_DISPLAY = "Table Blacklist";
 
+  public static final String SCHEMA_PATTERN_CONFIG = "schema.pattern";
+  private static final String SCHEMA_PATTERN_DOC =
+      "Schema pattern to fetch tables metadata from:\n"
+      + "  * \"\" retrieves those without a schema,"
+      + "  * null (default) means that the schema name should not be used to narrow the search, all tables "
+      + "metadata would be fetched, regardless there schema.";
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: typo: `s/there/their`

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 8]
```diff
@@ -108,6 +108,14 @@
   public static final String TABLE_BLACKLIST_DEFAULT = "";
   private static final String TABLE_BLACKLIST_DISPLAY = "Table Blacklist";
 
+  public static final String SCHEMA_PATTERN_CONFIG = "schema.pattern";
+  private static final String SCHEMA_PATTERN_DOC =
+      "Schema pattern to fetch tables metadata from:\n"
```
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: "from the database"

**Comment by shikhar** [comment]
@pqueixalos a few minor comments about the doc, otherwise LGTM

**Comment by shikhar** [comment]
Thanks for the contrib @pqueixalos!
